src/write/unit.rs

```diff
@@ -0,0 +1,1914 @@
+    /// Iterate over the attributes of this entry for modification.
+    #[inline]
+    pub fn attrs_mut(&mut self) -> slice::IterMut<Attribute> {
+        self.attrs.iter_mut()
+    }
+
+    /// Get an attribute.
+    pub fn get(&self, name: constants::DwAt) -> Option<&AttributeValue> {
+        self.attrs
+            .iter()
+            .find(|attr| attr.name == name)
+            .map(|attr| &attr.value)
+    }
+
+    /// Get an attribute for modification.
+    pub fn get_mut(&mut self, name: constants::DwAt) -> Option<&mut AttributeValue> {
+        self.attrs
+            .iter_mut()
+            .find(|attr| attr.name == name)
+            .map(|attr| &mut attr.value)
+    }
+
+    /// Set an attribute.
+    ///
+    /// Replaces any existing attribute with the same name.
+    pub fn set(&mut self, name: constants::DwAt, value: AttributeValue) {
+        if let Some(attr) = self.attrs.iter_mut().find(|attr| attr.name == name) {
+            attr.value = value;
+            return;
+        }
+        self.attrs.push(Attribute { name, value });
+    }
+
+    /// Iterate over the children of this entry.
+    ///
+    /// Note: use `CompilationUnit::add` to add a new child to this entry.
+    #[inline]
+    pub fn children(&self) -> slice::Iter<UnitEntryId> {
+        self.children.iter()
+    }
+
+    /// Return the type abbreviation for this DIE.
+    fn abbreviation(&self, format: Format) -> Result<Abbreviation> {
+        let mut attrs = Vec::new();
+        /* TODO: implement sibling support
+        if self.sibling {
+            let form = match format {
+                Format::Dwarf32 => constants::DW_FORM_ref4,
+                Format::Dwarf64 => constants::DW_FORM_ref8,
+            };
+            attrs.push(AttributeSpecification::new(constants::DW_AT_sibling, form));
+        }
+        */
+        for attr in &self.attrs {
+            attrs.push(attr.specification(format)?);
+        }
+        Ok(Abbreviation::new(
+            self.tag,
+            !self.children.is_empty(),
+            attrs,
+        ))
+    }
+
+    /// Write the entry to the given sections.
+    fn write<W: Writer>(
+        &self,
+        w: &mut DebugInfo<W>,
+        unit: &CompilationUnit,
+        offsets: &mut UnitOffsets,
+        abbrevs: &mut AbbreviationTable,
+        strings: &DebugStrOffsets,
+        unit_refs: &mut Vec<(DebugInfoOffset, UnitEntryId)>,
+        debug_info_refs: &mut Vec<(DebugInfoOffset, (UnitId, UnitEntryId), u8)>,
+    ) -> Result<()> {
+        offsets.entries[self.id.0] = w.offset();
+        let code = abbrevs.add(self.abbreviation(unit.format)?);
+        w.write_uleb128(code)?;
+        /* TODO: implement sibling support
+        if self.sibling {
+        }
+        */
+        for attr in &self.attrs {
+            attr.write(w, unit, strings, unit_refs, debug_info_refs)?;
+        }
+        if !self.children.is_empty() {
+            for child in &self.children {
+                unit.entries[child.0].write(
```

Comment:
It *might* be worth constructing the tree relationship eagerly once in `CompilationUnitHeader::write` or maintain it constantly while adding new DIEs. Because the DIEs can be mutated after creation and aren't append only, someone could (presumably accidentally) create DAGs and even cycles in the DIE "tree".

This creates two problems for us:

1. If there is a cycle, then we will blow the stack here because of the recursive implementation.

2. For a DAG, we can maybe generate bad debug info (not actually totally sure if this implementation would create copies of shared tails of the graph or emit bad debug info -- either way, it seems like something we probably want to catch and handle early and notify the user that they are doing something buggy)

Reply:
Currently we maintain the tree relationship constantly while adding new DIEs. Parents are specified when adding, and mutation doesn't allow changing parents. I'll add doc to make this clear.